airgun/session.py

```diff
@@ -161,24 +162,88 @@ def __init__(self, session_name=None, user=None, password=None, session_cookie=N
             passed - default one from settings will be used.
         :param requests.sessions.Session optional session_cookie: session object to be used
             to bypass login
+        :param str optional url: URL path to open when starting session (without protocol
+            and hostname)
         """
         self.name = session_name or gen_string('alphanumeric')
         self._user = user or settings.satellite.username
         self._password = password or settings.satellite.password
         self._session_cookie = session_cookie
         self._factory = None
+        self._url = None
+        self.navigator = None
         self.browser = None
 
+    def __call__(self, user=None, password=None, session_cookie=None, url=None):
+        """Stores provided values. This allows tests to provide additional
+        value when Session object is returned from fixture and used as
+        context manager. Arguments are the same as when initializing
+        Session object, except session_name.
+        """
+        if user is not None:
+            self._user = user
+        if password is not None:
+            self._password = password
+        if session_cookie is not None:
+            self._session_cookie = session_cookie
+        if url is not None:
+            self._url = url
+        return self
+
     def __enter__(self):
+        """Just a shim to make it compatible with context manager
+        protocol. The real work is done by _open the first time
+        any entity is requested.
+        """
+        return self
+
+    def __exit__(self, exc_type, exc_value, traceback):
+        """Attempts to log out, saves the screenshot and performs all required
+        session closure activities.
+
+        NOTE: exceptions during logout or saving screenshot are just logged and
+            not risen not to shadow real session result.
+        """
+        if self.browser is None:
+            # browser was never started, don't do anything
+            return
+        LOGGER.info(
+            u'Stopping UI session %r for user %r', self.name, self._user)
+        passed = True if exc_type is None else False
+        try:
+            if passed:
+                self.login.logout()
+            else:
+                self.take_screenshot()
+        except Exception as err:
+            LOGGER.exception(err)
+        finally:
+            self._factory.finalize(passed)
+
+    def _open(self, entity):
+        """Initializes requested entity. If this is first time session
+        requests an entity, also initialize and prepare browser.
+        """
+        if self.browser is None:
+            if self._url:
+                endpoint = self._url
+            else:
+                endpoint = getattr(entity, 'endpoint_path', '/')
+            full_url = f"https://{settings.satellite.hostname}{endpoint}"
```

Comment:
@mirekdlugosz do you think we should add the logger to know which path url being used?

Reply:
Sure, why not. Added in 2c0665beaf409c8d7a02527760f65e9da66643a5